Initial version of apiserver pod

luksa opened the pull request:
You may need to modify the name of the apiserver image in the pod YAML if you're not actually pushing the image to the registry used by k8s.

luksa commented on contrib/examples/apiserver/apiserver.yaml:
```diff
@@ -0,0 +1,34 @@
+    - name: https
+      containerPort: 6443
+  - name: etcd
+    image: gcr.io/google_containers/etcd:3.0.14
```
Unsure about which version to use. This one seems to work.

MHBauer replied:
What did k8s core _"support"_ in December? Where will it be at 1.6?

 I have 3.0.9 on one machine, 3.0.7 on another, and the integration test script use `quay.io/coreos/etcd`. @duglin 

I think it's fine.

pmorie replied:
@MHBauer not sure exactly what the question is here -- you're asking about what version k8s supports for a backing etcd?

MHBauer replied:
@pmorie I think that's what @luksa is asking. I was saying what we did have using. The quay image appears to be 3.1.0

duglin replied:
Would it make more sense to use the image from coreos instead?

duglin commented on contrib/examples/apiserver/apiserver.yaml:
```diff
@@ -0,0 +1,35 @@
+      containerPort: 6443
+  - name: etcd
+    image: gcr.io/google_containers/etcd:3.0.14
+    command: ["etcd"]
```
isn't the cmd defaulted in the image? If so, I'd prefer to leave it out.

luksa replied:
No, the `gcr.io/google_cotnainers/etcd` image simply runs `sh` by default. The `quay.io/coreos/etcd:v3.0.17`, on the other hand, does run `etcd`. Should I just use that one instead? What are the guidelines regarding the use of external images?

duglin commented on contrib/examples/apiserver/apiserver.yaml:
```diff
@@ -0,0 +1,35 @@
+    app: apiserver
+  ports:
+  - protocol: TCP
+    port: 6443
```
any reason we're not using 443?

luksa replied:
Two reasons:
- the [api-server-walkthrough](https://github.com/kubernetes-incubator/service-catalog/blob/master/docs/api-server-walkthrough.md) uses 6443 
- the k8s api server usually uses 8443, so 6443 makes it aligned with that

luksa commented:
Added instructions, used different etcd image, rebased to master.

arschles commented on contrib/examples/apiserver/apiserver.yaml:
```diff
@@ -0,0 +1,34 @@
+  - protocol: TCP
+    port: 6443
+    targetPort: https
+
```
let's also put a helm-compatible version of this into `deploy/catalog`

luksa replied:
done

MHBauer commented:
something does run. Going to need to adjust some ports given that 6443 is the default apiserver port.

MHBauer commented:
rebase to pull in the changes to #365 and adjust accordingly.

luksa commented:
Rebased, changed apiserver cert path, added apiserver to helm chart.

Haven't changed the port yet: if we really need to change it to something other than 6443, it needs to be done across the whole project (including changing the default in the catalog apiserver binary itself).

arschles commented:
thanks @luksa - LGTM

MHBauer commented on docs/api-server-walkthrough.md:
````diff
@@ -32,6 +32,16 @@ $ etcd
+Alternatively, you can run the apiserver and etcd as a pod:
+```
+# run the apiserver & etcd as a pod
+$ kubectl create -f contrib/examples/apiserver/apiserver.yaml
````
do we need to have run `make images` or the equivalent before this?

MHBauer commented on docs/api-server-walkthrough.md:
```diff
@@ -32,6 +32,16 @@ $ etcd
+# run the apiserver & etcd as a pod
+$ kubectl create -f contrib/examples/apiserver/apiserver.yaml
+# enable port forwarding from localhost:6443 to the apiserver
+$ kubectl port-forward service-catalog-apiserver 6443:6443
```
should be clear that this forwarder will block the terminal and we'll need to start a new one.

MHBauer replied:
this might be a pre-req for `hack/local-up-cluster.sh`, but you'll need `socat` and `nsenter` which were not installed for me. 

still trying to find nsenter.

MHBauer replied:
this should be an immediate check in kubectl if it relies on the two tools.

MHBauer commented on docs/api-server-walkthrough.md:
````diff
@@ -32,6 +32,16 @@ $ etcd
+```
+# run the apiserver & etcd as a pod
+$ kubectl create -f contrib/examples/apiserver/apiserver.yaml
+# enable port forwarding from localhost:6443 to the apiserver
````
wrt the port, possible warning that if you are running k8s on 6443, the port needs to be adjusted.

MHBauer commented:
If you can adjust the docs in the apiserver, that's good. If not, I need to go through and adjust the later section.

I went through the example and could do the apiserver operations and they worked correctly. Someone  needs to show me how to `helm` at some point.

lgtm

MHBauer commented:
I think mentioning `make images` is probably a hard requirement for change.

pmorie commented:
> If you can adjust the docs in the apiserver, that's good. If not, I need to go through and adjust the later section.

@luksa is going to be working on a broker going forward; I'll make an update PR to the docs, since I'm going to building on this tomorrow.

pmorie commented:
@luksa Thanks for the PR!
